File: src/Web/src/app/features/places/utils/prohibited-place-terms/prohibited-place-terms-catalog.factory.ts
import { CatalanProhibitedPlaceTermsCatalog } from './catalan-prohibited-place-terms.catalog';
import type { ProhibitedPlaceTermsCatalog } from './catalan-prohibited-place-terms.catalog';

const DEFAULT_LANGUAGE = 'ca';

/** Factory: add a language catalog here; the name filter consumes CreateAll. */
export class ProhibitedPlaceTermsCatalogFactory {
  constructor(private readonly catalogs: readonly ProhibitedPlaceTermsCatalog[]) {}

  create(languageCode = DEFAULT_LANGUAGE): ProhibitedPlaceTermsCatalog {
    const key = languageCode.trim().toLowerCase() || DEFAULT_LANGUAGE;
    const found = this.catalogs.find((catalog) => catalog.languageCode === key);
    if (found) {
      return found;
    }

    const catalan = this.catalogs.find((catalog) => catalog.languageCode === DEFAULT_LANGUAGE);
    if (catalan) {
      return catalan;
    }

    throw new Error('No prohibited-terms catalog is registered.');
  }

  createAll(): readonly ProhibitedPlaceTermsCatalog[] {
    return this.catalogs;
  }
}

export const prohibitedPlaceTermsCatalogFactory = new ProhibitedPlaceTermsCatalogFactory([
  new CatalanProhibitedPlaceTermsCatalog()
]);

File: src/Backend/Domain/Places/ProhibitedTerms/ProhibitedPlaceTermsCatalogFactory.cs
namespace Zuppeto.Domain.Places.ProhibitedTerms;

/// <summary>
/// Factory: new language = new <see cref="IProhibitedPlaceTermsCatalog"/> registration, no filter change.
/// </summary>
public sealed class ProhibitedPlaceTermsCatalogFactory : IProhibitedPlaceTermsCatalogFactory
{
    public const string DefaultLanguageCode = "ca";

    private readonly IReadOnlyDictionary<string, IProhibitedPlaceTermsCatalog> catalogsByLanguage;

    public ProhibitedPlaceTermsCatalogFactory(IEnumerable<IProhibitedPlaceTermsCatalog> catalogs)
    {
        var list = catalogs.ToArray();
        if (list.Length == 0)
        {
            list = [new CatalanProhibitedPlaceTermsCatalog()];
        }

        catalogsByLanguage = list.ToDictionary(
            catalog => catalog.LanguageCode,
            catalog => catalog,
            StringComparer.OrdinalIgnoreCase);
    }

    public IProhibitedPlaceTermsCatalog Create(string? languageCode)
    {
        var key = string.IsNullOrWhiteSpace(languageCode)
            ? DefaultLanguageCode
            : languageCode.Trim();

        if (catalogsByLanguage.TryGetValue(key, out var catalog))
        {
            return catalog;
        }

        return catalogsByLanguage.TryGetValue(DefaultLanguageCode, out var catalan)
            ? catalan
            : catalogsByLanguage.Values.First();
    }

    public IReadOnlyCollection<IProhibitedPlaceTermsCatalog> CreateAll() =>
        catalogsByLanguage.Values.ToArray();
}
